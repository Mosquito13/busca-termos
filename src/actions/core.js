import axios from 'axios';

import actionTypes from '../constants/actionTypes';
import releaseUtils from '../utils/releaseUtils';
import coreSelectors from '../selectors/core';
import apiUtils from '../utils/apiUtils';
import url from '../constants/url';

export const setLoading = (loading) => ({
  type: actionTypes.SET_LOADING,
  loading
});

export const setData = (data) => ({
  type: actionTypes.SET_DATA,
  data
});

export const setSelectedItemId = (id) => ({
  type: actionTypes.SET_SELECTED_ITEM_ID,
  id
});

export const setFilter = (filter) => ({
  type: actionTypes.SET_FILTER,
  filter
});

export const setAppVersion = (version) => ({
  type: actionTypes.SET_APP_VERSION,
  version
});

export const setAppHasUpdate = (appHasUpdate, updateURL) => ({
  type: actionTypes.SET_APP_HAS_UPDATE,
  appHasUpdate,
  updateURL
});

export const checkForUpdates =
  (currentVersion) => async (dispatch, getState) => {
    const state = getState();
    const hasUpdate = coreSelectors.getAppHasUpdate(state);

    if (hasUpdate === null) {
      const releases = await axios.get(url.LIST_APP_RELEASES);
      const [latestReleaseVersion, latestReleaseURL] =
        releaseUtils.parseLastRelease(releases);

      const hasUpdate = releaseUtils.hasUpdate(
        currentVersion,
        latestReleaseVersion
      );

      dispatch(setAppHasUpdate(hasUpdate, latestReleaseURL));
    }
  };

export const loadData = (settings) => async (dispatch) => {
  const data = await apiUtils.loadData(settings?.languageFolder);
  const appVersion = await apiUtils.loadAppVersion();

  dispatch(setAppVersion(appVersion));
  dispatch(checkForUpdates(appVersion));
  dispatch(setData(data));
  dispatch(setLoading(false));
};

export const openBrowserWithURL = (url) => async () => {
  apiUtils.openBrowserWithURL(url);
};
