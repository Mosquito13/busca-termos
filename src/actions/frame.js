import actionTypes from '../constants/actionTypes';
import apiUtils from '../utils/apiUtils';

export const setMaximized = (value) => ({
  type: actionTypes.SET_MAXIMIZED,
  value
});

export const close = () => () => apiUtils.close();

export const minimize = () => () => apiUtils.minimize();

export const maximize = () => () => apiUtils.maximize();

export const unmaximize = () => () => apiUtils.unmaximize();

export const registerFrameListeners = () => (dispatch) => {
  apiUtils.registerMaximizeListener(() => {
    dispatch(setMaximized(true));
  });

  apiUtils.registerUnmaximizeListener(() => {
    dispatch(setMaximized(false));
  });
};
