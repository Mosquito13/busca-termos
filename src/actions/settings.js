import apiUtils from '../utils/apiUtils';
import actionTypes from '../constants/actionTypes';
import storageUtils from '../utils/storageUtils';
import coreActions from './core';

const settingsActions = {
  setValid(value) {
    return {
      type: actionTypes.SET_VALID,
      value
    };
  },

  setValidating(value) {
    return {
      type: actionTypes.SET_VALIDATING,
      value
    };
  },

  setLanguageFolder(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER,
      value
    };
  },

  setLanguageFolderError(value) {
    return {
      type: actionTypes.SET_LANGUAGE_FOLDER_ERROR,
      value
    };
  },

  setMainLanguage(value) {
    return {
      type: actionTypes.SET_MAIN_LANGUAGE,
      value
    };
  },

  toggleTranslation(id, value) {
    return {
      type: actionTypes.TOGGLE_TRANSLATION,
      id,
      value
    };
  },

  toggleCompactLayout(value) {
    return {
      type: actionTypes.TOGGLE_COMPACT_LAYOUT,
      value
    };
  },

  toggleDarkTheme(value) {
    return {
      type: actionTypes.TOGGLE_DARK_THEME,
      value
    };
  },

  setSettings(value) {
    return {
      type: actionTypes.SET_SETTINGS,
      value
    };
  },

  validateAndSaveFirstSettings(languageFolder) {
    return async (dispatch) => {
      dispatch(this.setValidating(true));

      const isValid = await apiUtils.validateLanguageFolder(languageFolder);

      if (!isValid) {
        dispatch(this.setLanguageFolderError('Caminho invÃ¡lido.'));
        dispatch(this.setValid(false));
      } else {
        dispatch(coreActions.setLoading(true));
        dispatch(this.setValid(true));
        dispatch(this.saveSettings({ languageFolder }));
      }

      dispatch(this.setValidating(false));
    };
  },

  loadSettings() {
    return async (dispatch) => {
      const settings = storageUtils.getSettings();
      const isValid = await apiUtils.validateLanguageFolder(settings?.languageFolder);

      if (!isValid) {
        dispatch(this.setValid(false));
        dispatch(coreActions.setLoading(false));
      } else {
        dispatch(this.setSettings(settings));
        dispatch(coreActions.loadData(settings));
      }
    };
  },

  saveSettings(values) {
    return {
      type: actionTypes.SAVE,
      values
    };
  }
};

export default settingsActions;
