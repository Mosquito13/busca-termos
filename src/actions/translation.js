import actionTypes from '../constants/actionTypes';
import coreSelectors from '../selectors/core';

export const setTranslation = (id, value) => ({
  type: actionTypes.SET_TRANSLATION,
  id,
  value
});

export const loadTranslation = (id, termId) => (dispatch, getState) => {
  const state = getState();
  const data = coreSelectors.getData(state);
  const translation = data[id].find((d) => d.id === termId);

  dispatch(setTranslation(id, translation?.content));
};
