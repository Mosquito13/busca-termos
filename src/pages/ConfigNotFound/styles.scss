.config-not-found {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  &__message {
    text-align: center;
    width: 70%;
    display: flex;
    flex-direction: column;
    margin-top: 2rem;
  }

  &__message-title {
    margin-bottom: 2rem;
  }

  &__message-help {
    margin-bottom: 1rem;
  }

  &__message-help-example {
    margin-bottom: 1rem;
  }

  &__folder-picker {
    margin-top: 2rem;
    width: 50%;
  }

  &__btn-validate {
    margin-top: 2rem;
  }
}
