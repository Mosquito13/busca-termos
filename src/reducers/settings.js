import actionTypes from '../constants/actionTypes';
import storageUtils from '../utils/storageUtils';
import { languageMapping } from '../mapping/languages';

const translationInitial = {};

Object.values(languageMapping).forEach(({ id }) => translationInitial[id] = true);

const initial = {
  /* internal controls */
  valid: true,
  validating: false,
  languageFolderError: '',
  /* saved settings */
  languageFolder: '',
  mainLanguage: languageMapping.BRAZIL.id,
  translation: translationInitial,
  compactLayout: false,
  darkTheme: true
};

const getNewSettings = (state, values) => {
  const { languageFolder, mainLanguage, translation, compactLayout, darkTheme } = state;
  const currentSettings = {
    languageFolder,
    compactLayout,
    mainLanguage,
    translation,
    darkTheme
  };

  return {
    ...currentSettings,
    ...values
  };
};

const settingsReducer = (state = initial, action) => {
  switch (action.type) {
    case actionTypes.SET_VALID:
      return { ...state, valid: action.value };

    case actionTypes.SET_VALIDATING:
      return { ...state, validating: action.value };

    case actionTypes.SET_MAIN_LANGUAGE:
      return { ...state, mainLanguage: action.value };

    case actionTypes.TOGGLE_COMPACT_LAYOUT:
      return { ...state, compactLayout: action.value };

    case actionTypes.TOGGLE_DARK_THEME:
      return { ...state, darkTheme: action.value };

    case actionTypes.TOGGLE_TRANSLATION:
      return {
        ...state,
        translation: {
          ...state.translation,
          [action.id]: action.value
        }
      };

    case actionTypes.SET_LANGUAGE_FOLDER:
      return { ...state, languageFolder: action.value };

    case actionTypes.SET_LANGUAGE_FOLDER_ERROR:
      return { ...state, languageFolderError: action.value };

    case actionTypes.SET_SETTINGS:
      return { ...state, ...getNewSettings(state, action.value) };

    case actionTypes.SAVE:
      storageUtils.setSettings(getNewSettings(state, action.values));
      return { ...state, languageFolderError: '' };

    default:
      return state;
  }
};

export default settingsReducer;
