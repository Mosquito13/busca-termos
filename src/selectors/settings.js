const getSlice = (state) => state.Settings;

const settingsSelectors = {
  isValid(state) {
    return getSlice(state).valid;
  },

  isValidating(state) {
    return getSlice(state).validating;
  },

  isDarkTheme(state) {
    return getSlice(state).darkTheme;
  },

  isCompactLayout(state) {
    return getSlice(state).compactLayout;
  },

  getMainLanguage(state) {
    return getSlice(state).mainLanguage;
  },

  getTranslation(state) {
    return getSlice(state).translation;
  },

  getLanguageFolder(state) {
    return getSlice(state).languageFolder;
  },

  getLanguageFolderError(state) {
    return getSlice(state).languageFolderError;
  }
};

export default settingsSelectors;
