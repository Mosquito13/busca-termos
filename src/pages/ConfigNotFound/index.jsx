import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FiAlertTriangle } from 'react-icons/fi';
import { useNavigate } from 'react-router-dom';

import Button from '../../components/common/Button';
import Heading3 from '../../components/common/Heading3';
import Paragraph from '../../components/common/Paragraph';
import Icon, { LARGE } from '../../components/common/Icon';
import LanguageFolderPicker from '../../components/LanguageFolderPicker';

import { setLanguageFolder, validateAndSaveFirstSettings } from '../../actions/settings';
import settingsSelectors from '../../selectors/settings';

import './styles.scss';

const ConfigNotFound = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const isValid = useSelector(settingsSelectors.isValid);
  const isValidating = useSelector(settingsSelectors.isValidating);
  const languageFolder = useSelector(settingsSelectors.getLanguageFolder);
  const languageFolderError = useSelector(settingsSelectors.getLanguageFolderError);

  useEffect(() => {
    if (isValid) {
      navigate('/');
    }
  }, [isValid, navigate]);

  const handleChangeLanguageFolder = useCallback((value) => {
    dispatch(setLanguageFolder(value));
  }, [dispatch]);

  const onClickValidate = useCallback(() => {
    dispatch(validateAndSaveFirstSettings(languageFolder));
  }, [dispatch, languageFolder]);

  return (
    <div className="config-not-found">
      <Icon size={LARGE}>
        <FiAlertTriangle />
      </Icon>
      <div className="config-not-found__message">
        <div className="config-not-found__message-title">
          <Heading3 value="Configuração não encontrada." />
        </div>
        <div className="config-not-found__message-help">
          <Paragraph value="Insira ou selecione o caminho até a pasta language do seu repositório." />
        </div>
        <div className="config-not-found__message-help-example">
          <Paragraph value ="A pasta language fica em {{seu_repositorio}}/System/web/include/language" smallFont />
        </div>
      </div>
      <div className="config-not-found__folder-picker">
        <LanguageFolderPicker
          value={languageFolder}
          error={languageFolderError}
          onChange={handleChangeLanguageFolder}
          disabled={isValidating}
          data-testid="folder-picker"
        />
      </div>
      <div className="config-not-found__btn-validate">
        <Button
          text="Validar e salvar"
          loading={isValidating}
          onClick={onClickValidate}
          data-testid="btn-save"
        />
      </div>
    </div>
  );
};

export default ConfigNotFound;
