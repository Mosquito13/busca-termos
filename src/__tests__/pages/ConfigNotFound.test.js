import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';

import ConfigNotFound from '../../pages/ConfigNotFound';
import * as settingsActions from '../../actions/settings';

const navigateMock = jest.fn();
const reactRouterDomMock = { useNavigate: () => navigateMock };

jest.mock('react-router-dom', () => reactRouterDomMock);

const getStateMock = (valid, languageFolder) => ({
  Settings: {
    valid,
    validating: false,
    languageFolder
  }
});

const prepareMocks = (dispatchMock, stateMock) => {
  jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock);
  jest
    .spyOn(reactRedux, 'useSelector')
    .mockImplementation((selector) => selector(stateMock));
};

describe(`${ConfigNotFound.name}`, () => {
  describe('tests', () => {
    let wrapper, dispatchMock;

    beforeAll(() => {
      dispatchMock = jest.fn();

      prepareMocks(dispatchMock, getStateMock(false));

      jest.spyOn(settingsActions, 'setLanguageFolder');
      jest.spyOn(settingsActions, 'validateAndSaveFirstSettings');

      wrapper = mount(<ConfigNotFound />);
    });

    it('should render and stay on page while folder is not valid', () => {
      expect(wrapper.find('.config-not-found')).toHaveLength(1);
      expect(navigateMock).toBeCalledTimes(0);
    });

    it('should set language folder', () => {
      wrapper
        .find('[data-testid="folder-picker"]')
        .props()
        .onChange('a/folder/');

      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(
        settingsActions.setLanguageFolder('a/folder/')
      );
    });

    it('should validate when click save', () => {
      wrapper.unmount();

      prepareMocks(dispatchMock, getStateMock(false, 'a/folder/that/is/set/'));

      wrapper = mount(<ConfigNotFound />);

      wrapper.find('[data-testid="btn-save"]').props().onClick();

      expect(dispatchMock).toBeCalledTimes(1);
      expect(dispatchMock).toBeCalledWith(
        settingsActions.validateAndSaveFirstSettings('a/folder/that/is/set/')
      );
    });

    it('should navigate to main page when the folder is valid', () => {
      wrapper.unmount();

      prepareMocks(dispatchMock, getStateMock(true));

      wrapper = mount(<ConfigNotFound />);

      expect(navigateMock).toBeCalledTimes(1);
      expect(navigateMock).toBeCalledWith('/');
    });
  });

  describe('snapshots', () => {
    it('should render', () => {
      prepareMocks(jest.fn(), getStateMock(false));

      const tree = renderer.create(<ConfigNotFound />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
