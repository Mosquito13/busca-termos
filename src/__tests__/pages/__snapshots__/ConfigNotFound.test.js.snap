// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigNotFound snapshots should render 1`] = `
<div
  className="config-not-found"
>
  <div
    className="icon icon--large"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"
      />
      <line
        x1="12"
        x2="12"
        y1="9"
        y2="13"
      />
      <line
        x1="12"
        x2="12.01"
        y1="17"
        y2="17"
      />
    </svg>
  </div>
  <div
    className="config-not-found__message"
  >
    <div
      className="config-not-found__message-title"
    >
      <h3
        className="heading-3"
      >
        Configuração não encontrada.
      </h3>
    </div>
    <div
      className="config-not-found__message-help"
    >
      <p
        className="paragraph"
      >
        Insira ou selecione o caminho até a pasta language do seu repositório.
      </p>
    </div>
    <div
      className="config-not-found__message-help-example"
    >
      <p
        className="paragraph paragraph--small"
      >
        A pasta language fica em {{seu_repositorio}}/System/web/include/language
      </p>
    </div>
  </div>
  <div
    className="config-not-found__folder-picker"
  >
    <div
      className="language-folder-picker"
    >
      <div
        className="language-folder-picker__fields"
      >
        <div
          className="input-text"
        >
          <input
            className="input-text__input"
            disabled={false}
            onChange={[Function]}
            placeholder="Caminho até a pasta language..."
            spellCheck={false}
            type="text"
            value=""
          />
        </div>
        <div
          className="language-folder-picker__button"
        >
          <button
            className="button button--with-icon"
            disabled={false}
            onClick={[Function]}
            title="Selecionar pasta..."
          >
            <div
              className="button__content"
            >
              <div
                className="icon icon--medium"
              >
                <svg
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="1"
                  />
                  <circle
                    cx="19"
                    cy="12"
                    r="1"
                  />
                  <circle
                    cx="5"
                    cy="12"
                    r="1"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
        <input
          className="language-folder-picker__input-file"
          directory=""
          id="filepicker"
          name="fileList"
          onChange={[Function]}
          type="file"
          webkitdirectory=""
        />
      </div>
      <div
        className="language-folder-picker__error language-folder-picker__error--hidden"
      >
        
      </div>
    </div>
  </div>
  <div
    className="config-not-found__btn-validate"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      <div
        className="button__content"
      >
        Validar e salvar
      </div>
    </button>
  </div>
</div>
`;
